### DISCUSSION TITLE
Page get previous props.

### DISCUSSION THREAD

**Discussion by foxter111** [vercel/next.js#39973] [Help] [answered]
Hello everyone. I have a problem with my page props. Sometimes, my page get previous props, although I see, that getServerSideProps passes correct values.

I have a page that render different content depending on the type of user. I get type of user in getUserTypeServer and pass it to the page. 

**My code for example:**

```
 const Page = props => {
    if (props.userType === 'admin') {
        return <AdminPage />;
    }
    return <UserPage />;
};

export async function getServerSideProps(ctx) {
    const { req } = ctx;
    const headers = headersForServer(req);
    let userType = null;

    if (headers['x-auth-token']) userType = await getUserTypeServer(headers);

    return {
        props: {
            userType,
        },
    };
}

export default Page;
```

**My actions:**
1) Log in for user
2) Go to the page
3) Logging out
4) Log in for admin
5) Go to the page

My page gets previous props and renders a component for ordinary user, although getUserTypeServer passes a correct type.
After reload page, it gets correct props and renders a component for admin.  I think, it can be caching, but I don't know, how I can fix it. Does somebody help me? I'm not sure, but I seem, that it can be pwa.

**My next.config:**

```
const moduleExports = withPWA({
    pwa: {
        dest: 'public',
        register: true,
        skipWaiting: true,
        disable: process.env.NODE_ENV === 'development',
    },
    compiler: {
        styledComponents: true,
    },
    swcMinify: true,
    async redirects() {
        return [
            {
                source: '/',
                destination: '/login',
                permanent: false,
            },
        ];
    },
    webpack: config => {
        config.resolve.alias['@'] = path.resolve(__dirname, 'src');
        return config;
    },
});

module.exports = moduleExports;
```

**Version libraries:**
"next": "12.2.2",
 "next-pwa": "5.5.4",

**Comment by icyJoseph** [comment, marked as answer]
If you think this is because of a PWA, I'd review the Service Worker Strategy.

It is probable that the Service Worker is doing the stale-while-revalidate strategy, https://developer.chrome.com/docs/workbox/modules/workbox-strategies/, which is most likely not what you want in this case. 

I guess in your case the Network First (Network Falling Back to Cache), it's what's most appropriate, but `next-pwa` has stale-while-revalidate as default.

```js
/** @type {import('next').NextConfig} */
const withPWA = require("next-pwa");
const runtimeCaching = require("next-pwa/cache");

const nextDataIndex = runtimeCaching.findIndex(
  (entry) => entry.options.cacheName === "next-data"
);

if (nextDataIndex !== -1) {
  runtimeCaching[nextDataIndex].handler = "NetworkFirst";
} else {
  throw new Error("Failed to find next-data object in runtime caching");
}

// now runtimeCaching has been mutated
module.exports = withPWA({
  pwa: {
    dest: "public",
    runtimeCaching,
    register: true,
    skipWaiting: true,
    buildExcludes: [/middleware-manifest.json$/],
  }
});

```

**Reply by foxter111** [reply]
@icyJoseph Thank you so much. It works for me.

**Comment by foxter111** [comment]
Thank you. I will try it and report the results later.
